apis/s3/v1alpha2/referencers.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2020 The Crossplane Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/crossplane/crossplane-runtime/pkg/reference"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/provider-aws/apis/identity/v1alpha1"
+	identityv1beta1 "github.com/crossplane/provider-aws/apis/identity/v1beta1"
+	"github.com/crossplane/provider-aws/apis/s3/v1beta1"
+)
+
+// ResolveReferences of this BucketPolicy
+func (mg *BucketPolicy) ResolveReferences(ctx context.Context, c client.Reader) error {
+	r := reference.NewAPIResolver(c, mg)
+	// Resolve spec.forProvider.bucketName
+	rsp, err := r.Resolve(ctx, reference.ResolutionRequest{
+		CurrentValue: reference.FromPtrValue(mg.Spec.PolicyBody.BucketName),
+		Reference:    mg.Spec.PolicyBody.BucketNameRef,
+		Selector:     mg.Spec.PolicyBody.BucketNameSelector,
+		To:           reference.To{Managed: &v1beta1.Bucket{}, List: &v1beta1.BucketList{}},
+		Extract:      reference.ExternalName(),
+	})
+	if err != nil {
+		return errors.Wrap(err, "spec.forProvider.bucketName")
+	}
+	mg.Spec.PolicyBody.BucketName = reference.ToPtrValue(rsp.ResolvedValue)
+	mg.Spec.PolicyBody.BucketNameRef = rsp.ResolvedReference
+
+	// Resolve spec.forProvider.userName
+	if mg.Spec.PolicyBody.PolicyStatement != nil {
+		for i := range mg.Spec.PolicyBody.PolicyStatement {
+			statement := mg.Spec.PolicyBody.PolicyStatement[i]
+			err = ResolvePrincipal(ctx, r, statement.Principal, i)
+			if err != nil {
+				return err
+			}
+			err = ResolvePrincipal(ctx, r, statement.NotPrincipal, i)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// ResolvePrincipal resolves all the IAMUser and IAMRole references in a BucketPrincipal
+func ResolvePrincipal(ctx context.Context, r *reference.APIResolver, principal *BucketPrincipal, statementIndex int) error {
+	if principal == nil || principal.AWSPrincipals == nil {
```

Comment:
```suggestion
	if principal == nil {
```

If the array is nil, for loop would just exit and `return nil` would be executed.